scripts/utils.py

```diff
@@ -60,6 +60,83 @@ def rename_columns(
     return df
 
 
+def extract_doi(
+    df: pd.DataFrame,
+    col_in: str,
+    pre: str = "",
+    post: str = "",
+    overwrite: bool = False,
+):
+    """Extracts doi with regex taking into account pre and post. Stores it in doi column with doi.org prefix."""
+
+    regex = rf"{pre}(10\.[^{post}]*)" if post else rf"{pre}(10\.\S+)"
+    url_prefix = "https://doi.org/"
+
+    if ("doi" not in list(df)) or overwrite:
+        df["doi"] = url_prefix + df[col_in].str.extract(regex)[0]
+    else:
+        # Only overwrite rows that did not have a doi yet.
+        df["doi"] = df["doi"].mask(
+            df["doi"].isnull(), url_prefix + df[col_in].str.extract(regex)[0]
+        )
+
+    return df
+
+10\.[^&]*
+def extract_pmid(
+    df: pd.DataFrame,
+    col_in: str,
+    pre: str = "",
+    post: str = "",
+    overwrite: bool = False,
+):
+    """Extracts pmid with regex taking into account pre and post. Stores it in pmid column with pubmed prefix."""
+
+    regex = rf"{pre}(\d+[^{post}]*)" if post else rf"{pre}(\d+)"
+    url_prefix = "https://pubmed.ncbi.nlm.nih.gov/"
+
+    if ("pmid" not in list(df)) or overwrite:
+        df["pmid"] = url_prefix + df[col_in].str.extract(regex)[0]
+    else:
+        # Only overwrite rows that did not have a pmid yet.
+        df["pmid"] = df["pmid"].mask(
+            df["pmid"].isnull(), url_prefix + df[col_in].str.extract(regex)[0]
+        )
+
+    return df
+
+
+def extract_year(df: pd.DataFrame, col_in: str, overwrite: bool = False):
+    """Extracts the number from the input column and stores it in year column."""
+
+    regex = r"(\d+)"
+
+    if ("year" not in list(df)) or overwrite:
+        df["year"] = df[col_in].str.extract(regex)
+    elif df[col_in].dtype == "object":
```

Comment:
Better use this: `from pandas.api.types import is_string_dtype`. pandas is changing a lot for v3 on string columns

Reply:
Nice that they are working on strings